<a href="https://colab.research.google.com/github/rishabh-0042/American_Express_Hackathon_25-2025/blob/main/Feature_Engineering3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:

import pandas as pd

test_data = pd.read_parquet('/content/drive/MyDrive/AMEXChallenge/test_data.parquet')
train_data = pd.read_parquet('/content/drive/MyDrive/AMEXChallenge/train_data.parquet')

print("Test data loaded successfully.")
print("Train data loaded successfully.")

# Display the first few rows of each DataFrame to verify
print("\nTest data head:")
display(test_data.head())

print("\nTrain data head:")
display(train_data.head())

Test data loaded successfully.
Train data loaded successfully.

Test data head:


id1      id2     id3  \
46756    1362907_91950_16-23_2023-11-04 18:56:26.000794  1362907   91950   
57819       1082599_88356_16-23_2023-11-04 06:08:53.373  1082599   88356   
15390   1888466_958700_16-23_2023-11-05 10:07:28.000725  1888466  958700   
145730     1888971_795739_16-23_2023-11-04 12:25:28.244  1888971  795739   
146085      1256369_82296_16-23_2023-11-05 06:45:26.657  1256369   82296   

                               id4         id5    f1    f2    f3    f4    f5  \
46756   2023-11-04 18:56:26.000794  2023-11-04  None  None  None  None  None   
57819      2023-11-04 06:08:53.373  2023-11-04  None   9.0  None  None  None   
15390   2023-11-05 10:07:28.000725  2023-11-05  None  None  None  None  22.0   
145730     2023-11-04 12:25:28.244  2023-11-04  None  None  None  None  None   
146085     2023-11-05 06:45:26.657  2023-11-05  None  None  None  None  None   

        ...                 f357                f358 f359  f360   f361  f362  \
46756   ...   0.0027807272790246  0.0465999838670646  0.0  None    1.0   0.0   
57819   ...   0.0014287946777398  0.0603093514970909  0.0  None  195.0  13.0   
15390   ...  -0.0174961119751166  0.0734836702954899  0.0  None  155.0  67.0   
145730  ...   0.0013159476327096   0.040572039549215  0.0  None   None  None   
146085  ...   0.0020541699799278   0.038243539079214  0.0  None   29.0   2.0   

                      f363    f364   f365                f366  
46756                  0.0    56.0    0.0                 0.0  
57819   0.0666666666666666    None   None                None  
15390    0.432258064516129  1142.0  436.0  0.3817863397548161  
145730                None    None   None                None  
146085  0.0689655172413793   361.0    3.0  0.0083102493074792  

[5 rows x 371 columns]


Train data head:


id1      id2        id3  \
0  1366776_189706075_16-23_2023-11-02 22:22:00.042  1366776  189706075   
1      1366776_89227_16-23_2023-11-01 23:51:24.999  1366776      89227   
2      1366776_35046_16-23_2023-11-01 00:30:59.797  1366776      35046   
3    1366776_6275451_16-23_2023-11-02 22:21:32.261  1366776    6275451   
4      1366776_78053_16-23_2023-11-02 22:21:34.799  1366776      78053   

                       id4         id5  y   f1    f2    f3    f4  ...  f357  \
0  2023-11-02 22:22:00.042  2023-11-02  0  1.0  None  None  None  ...  None   
1  2023-11-01 23:51:24.999  2023-11-01  0  1.0  None  None  None  ...  None   
2  2023-11-01 00:30:59.797  2023-11-01  0  1.0  None  None  None  ...  None   
3  2023-11-02 22:21:32.261  2023-11-02  0  1.0  None  None  None  ...  None   
4  2023-11-02 22:21:34.799  2023-11-02  0  1.0  None  None  None  ...  None   

      f358 f359  f360   f361 f362               f363    f364 f365  \
0  -9999.0  0.0  None   28.0  0.0                0.0   337.0  0.0   
1     None  0.0  None   87.0  0.0                0.0  1010.0  2.0   
2     None  0.0  None   23.0  0.0                0.0  1010.0  2.0   
3  -9999.0  0.0  None  277.0  1.0  0.003610108303249   337.0  0.0   
4  -9999.0  0.0  None  359.0  0.0                0.0   337.0  0.0   

                 f366  
0                 0.0  
1  0.0019801980198019  
2  0.0019801980198019  
3                 0.0  
4                 0.0  

[5 rows x 372 columns]

In [ ]:
f325to328_train = train_data[['id2', 'f325', 'f326', 'f327', 'f328']]
f325to328_test = test_data[['id2', 'f325', 'f326', 'f327', 'f328']]

print("Train subset created successfully.")
display(f325to328_train.head())

print("\nTest subset created successfully.")
display(f325to328_test.head())

Train subset created successfully.


id2  f325  f326  f327  f328
0  1366776  None  None  None  None
1  1366776  None  None  None  None
2  1366776  None  None  None  None
3  1366776  None  None  None  None
4  1366776  None  None  None  None


Test subset created successfully.


id2                f325                f326                f327  \
46756   1362907  0.8792872910753705   1.040797370689633  0.9217916314743244   
57819   1082599  1.1682670324645634  1.0874633072407045  0.9577911072602342   
15390   1888466  0.8225943721954622                 1.0                 1.0   
145730  1888971  0.9435086146160252    1.01567338737897   0.962999694661668   
146085  1256369  0.8921319103848296  1.0834134185777535  0.9327473321862249   

                      f328  
46756   0.9267417375856436  
57819   0.8624490793734982  
15390                  1.0  
145730  0.9433644526349622  
146085  0.9196601678872748

In [ ]:
import numpy as np

def analyze_dataframe(df, name):
    print(f"Analyzing {name} DataFrame:")
    print("-" * 30)
    for col in df.columns:
        if df[col].dtype in [np.float64, np.int64]:
            null_percentage = df[col].isnull().sum() / len(df) * 100
            print(f"Column '{col}':")
            print(f"  Null Percentage: {null_percentage:.2f}%")
            data_range = f"{df[col].min():.2f} - {df[col].max():.2f}"
            print(f"  Range: {data_range}")
            print("-" * 20)
        else:
            print(f"Column '{col}':")
            print("  Not a numeric data type.")
            print("-" * 20)


analyze_dataframe(f325to328_train, "f325to328_train")
analyze_dataframe(f325to328_test, "f325to328_test")

Analyzing f325to328_train DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f325':
  Null Percentage: 25.42%
  Range: 0.00 - 18.69
--------------------
Column 'f326':
  Null Percentage: 25.42%
  Range: 0.00 - 26.92
--------------------
Column 'f327':
  Null Percentage: 25.41%
  Range: 0.00 - 30.92
--------------------
Column 'f328':
  Null Percentage: 25.41%
  Range: 0.19 - 37.87
--------------------
Analyzing f325to328_test DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f325':
  Null Percentage: 1.82%
  Range: 0.00 - 15.40
--------------------
Column 'f326':
  Null Percentage: 1.82%
  Range: 0.00 - 28.04
--------------------
Column 'f327':
  Null Percentage: 1.82%
  Range: 0.00 - 17.35
--------------------
Column 'f328':
  Null Percentage: 1.82%
  Range: 0.20 - 31.76
--------------------


In [ ]:
print("Data types for f325to328_train:")
print(f325to328_train.dtypes)

print("\nData types for f325to328_test:")
print(f325to328_test.dtypes)

Data types for f325to328_train:
id2     object
f325    object
f326    object
f327    object
f328    object
dtype: object

Data types for f325to328_test:
id2     object
f325    object
f326    object
f327    object
f328    object
dtype: object


In [ ]:
for col in ['f325', 'f326', 'f327', 'f328']:
    f325to328_train[col] = pd.to_numeric(f325to328_train[col], errors='coerce')
    f325to328_test[col] = pd.to_numeric(f325to328_test[col], errors='coerce')

print("Columns converted to float in f325to328_train:")
print(f325to328_train.dtypes)

print("\nColumns converted to float in f325to328_test:")
print(f325to328_test.dtypes)

print("\nFirst few rows of f325to328_train after conversion:")
display(f325to328_train.head())

print("\nFirst few rows of f325to328_test after conversion:")
display(f325to328_test.head())

/tmp/ipython-input-7-68738084.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f325to328_train[col] = pd.to_numeric(f325to328_train[col], errors='coerce')
/tmp/ipython-input-7-68738084.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f325to328_test[col] = pd.to_numeric(f325to328_test[col], errors='coerce')


Columns converted to float in f325to328_train:
id2      object
f325    float64
f326    float64
f327    float64
f328    float64
dtype: object

Columns converted to float in f325to328_test:
id2      object
f325    float64
f326    float64
f327    float64
f328    float64
dtype: object

First few rows of f325to328_train after conversion:


id2  f325  f326  f327  f328
0  1366776   NaN   NaN   NaN   NaN
1  1366776   NaN   NaN   NaN   NaN
2  1366776   NaN   NaN   NaN   NaN
3  1366776   NaN   NaN   NaN   NaN
4  1366776   NaN   NaN   NaN   NaN


First few rows of f325to328_test after conversion:


id2      f325      f326      f327      f328
46756   1362907  0.879287  1.040797  0.921792  0.926742
57819   1082599  1.168267  1.087463  0.957791  0.862449
15390   1888466  0.822594  1.000000  1.000000  1.000000
145730  1888971  0.943509  1.015673  0.963000  0.943364
146085  1256369  0.892132  1.083413  0.932747  0.919660

In [ ]:
from sklearn.preprocessing import MinMaxScaler
import numpy as np

# Apply log1p transformation
for col in ['f325', 'f326', 'f327', 'f328']:
    # Applying log1p to handle potential zeros and skewed data
    # Add a small constant or handle zeros if necessary before log1p
    # For now, assuming non-negative values for log1p
    f325to328_train[col] = np.log1p(f325to328_train[col])
    f325to328_test[col] = np.log1p(f325to328_test[col])


# Normalize the data using MinMaxScaler
scaler = MinMaxScaler()

# Fit the scaler on the training data and transform both train and test data
# Only apply to the numeric columns
numeric_cols = ['f325', 'f326', 'f327', 'f328']

f325to328_train[numeric_cols] = scaler.fit_transform(f325to328_train[numeric_cols])
f325to328_test[numeric_cols] = scaler.transform(f325to328_test[numeric_cols])


print("Log1p transformation and normalization applied.")

print("\nFirst few rows of f325to328_train after transformation and normalization:")
display(f325to328_train.head())

print("\nFirst few rows of f325to328_test after transformation and normalization:")
display(f325to328_test.head())

Log1p transformation and normalization applied.

First few rows of f325to328_train after transformation and normalization:


/tmp/ipython-input-9-3035108624.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f325to328_train[col] = np.log1p(f325to328_train[col])
/tmp/ipython-input-9-3035108624.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f325to328_test[col] = np.log1p(f325to328_test[col])
/tmp/ipython-input-9-3035108624.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pan

id2  f325  f326  f327  f328
0  1366776   NaN   NaN   NaN   NaN
1  1366776   NaN   NaN   NaN   NaN
2  1366776   NaN   NaN   NaN   NaN
3  1366776   NaN   NaN   NaN   NaN
4  1366776   NaN   NaN   NaN   NaN


First few rows of f325to328_test after transformation and normalization:


id2      f325      f326      f327      f328
46756   1362907  0.211702  0.214252  0.188624  0.138922
57819   1082599  0.259699  0.221043  0.193983  0.129195
15390   1888466  0.201424  0.208187  0.200142  0.149617
145730  1888971  0.222978  0.210532  0.194750  0.141384
146085  1256369  0.213988  0.220459  0.190266  0.137866

In [ ]:
# Define the columns to impute
cols_to_impute = ['f325', 'f326', 'f327', 'f328']

# Calculate the overall median for each column as a fallback
overall_medians_train = f325to328_train[cols_to_impute].median()
overall_medians_test = f325to328_test[cols_to_impute].median()

# Define a function to impute based on group median with overall median fallback
def impute_with_group_median(group, overall_medians):
    return group.fillna(group.median()).fillna(overall_medians)

# Apply the imputation to the training data
f325to328_train[cols_to_impute] = f325to328_train.groupby('id2')[cols_to_impute].transform(
    lambda x: impute_with_group_median(x, overall_medians_train)
)

# Apply the imputation to the test data
f325to328_test[cols_to_impute] = f325to328_test.groupby('id2')[cols_to_impute].transform(
    lambda x: impute_with_group_median(x, overall_medians_test)
)

print("Missing values imputed based on group median with overall median fallback.")

print("\nFirst few rows of f325to328_train after imputation:")
display(f325to328_train.head())

print("\nFirst few rows of f325to328_test after imputation:")
display(f325to328_test.head())

# Verify if there are any remaining null values in the imputed columns
print("\nNull value check after imputation for f325to328_train:")
print(f325to328_train[cols_to_impute].isnull().sum())

print("\nNull value check after imputation for f325to328_test:")
print(f325to328_test[cols_to_impute].isnull().sum())

/tmp/ipython-input-10-4044956794.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f325to328_train[cols_to_impute] = f325to328_train.groupby('id2')[cols_to_impute].transform(


Missing values imputed based on group median with overall median fallback.

First few rows of f325to328_train after imputation:


/tmp/ipython-input-10-4044956794.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f325to328_test[cols_to_impute] = f325to328_test.groupby('id2')[cols_to_impute].transform(


id2      f325      f326      f327      f328
0  1366776  0.231303  0.191361  0.190917  0.139014
1  1366776  0.231303  0.191361  0.190917  0.139014
2  1366776  0.231303  0.191361  0.190917  0.139014
3  1366776  0.231303  0.191361  0.190917  0.139014
4  1366776  0.231303  0.191361  0.190917  0.139014


First few rows of f325to328_test after imputation:


id2      f325      f326      f327      f328
46756   1362907  0.211702  0.214252  0.188624  0.138922
57819   1082599  0.259699  0.221043  0.193983  0.129195
15390   1888466  0.201424  0.208187  0.200142  0.149617
145730  1888971  0.222978  0.210532  0.194750  0.141384
146085  1256369  0.213988  0.220459  0.190266  0.137866


Null value check after imputation for f325to328_train:
f325    0
f326    0
f327    0
f328    0
dtype: int64

Null value check after imputation for f325to328_test:
f325    0
f326    0
f327    0
f328    0
dtype: int64


In [16]:
f325to328_train.to_parquet('/content/drive/MyDrive/Transformed_Train_Data.parquet')
f325to328_test.to_parquet('/content/drive/MyDrive/Transformed_Test_Data.parquet')

print("Transformed train data saved to /content/drive/MyDrive/Transformed_Train_Data.parquet")
print("Transformed test data saved to /content/drive/MyDrive/Transformed_Test_Data.parquet")

Transformed train data saved to /content/drive/MyDrive/Transformed_Train_Data.parquet
Transformed test data saved to /content/drive/MyDrive/Transformed_Test_Data.parquet


In [12]:
try:
    loaded_test_data = pd.read_parquet('/content/drive/MyDrive/Transformed_Test_Data.parquet')
    print("Successfully loaded the saved data.")
    print("\nFirst few rows of the loaded data:")
    display(loaded_test_data.head())
except FileNotFoundError:
    print("Error: The file was not found at the specified path.")
except Exception as e:
    print(f"An error occurred: {e}")

Successfully loaded the saved data.

First few rows of the loaded data:


id2      f325      f326      f327      f328
46756   1362907  0.211702  0.214252  0.188624  0.138922
57819   1082599  0.259699  0.221043  0.193983  0.129195
15390   1888466  0.201424  0.208187  0.200142  0.149617
145730  1888971  0.222978  0.210532  0.194750  0.141384
146085  1256369  0.213988  0.220459  0.190266  0.137866

In [14]:
!ls -l /content/drive/MyDrive/Transformed_Test_Data.parquet

-rw------- 1 root root 4981884 Jun 29 14:32 /content/drive/MyDrive/Transformed_Test_Data.parquet


In [15]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [20]:
import os
import pandas as pd

# Define the directory path
save_directory = '/content/drive/MyDrive/Transformed_Test_Data'

# Create the directory if it doesn't exist
os.makedirs(save_directory, exist_ok=True)

# Define the file path for saving the test DataFrame
file_path_save_test_325_328 = os.path.join(save_directory, 'f325to328_test_processed.parquet')

try:
    # Save the test DataFrame
    f325to328_test.to_parquet(file_path_save_test_325_328, index=False)
    print(f"Successfully saved the processed f325to328 test DataFrame to {file_path_save_test_325_328}")

except Exception as e:
    print(f"An error occurred while saving the DataFrame: {e}")

Successfully saved the processed f325to328 test DataFrame to /content/drive/MyDrive/Transformed_Test_Data/f325to328_test_processed.parquet


**Reasoning**:
Create subsets of the train and test dataframes with the specified columns and display their heads to verify.



In [21]:
f329to331_train = train_data[['id2', 'f329', 'f330', 'f331']]
f329to331_test = test_data[['id2', 'f329', 'f330', 'f331']]

print("Train subset created successfully.")
display(f329to331_train.head())

print("\nTest subset created successfully.")
display(f329to331_test.head())

Train subset created successfully.


id2  f329  f330  f331
0  1366776  None  None  None
1  1366776  None  None  None
2  1366776  None  None  None
3  1366776  None  None  None
4  1366776  None  None  None


Test subset created successfully.


id2                f329                f330                f331
46756   1362907   0.844820822801228  1.1291026465763951  0.9946585915896966
57819   1082599   1.074304783146098   1.135386723678609  1.1105480081861696
15390   1888466  0.8225943721954622                 1.0                 1.0
145730  1888971  0.9289488395977646  1.0546975175685886  1.0208140575700744
146085  1256369  0.8234455057386793  1.1615293672707585   1.014230435062785

**Reasoning**:
Define the `analyze_dataframe` function and call it for both the train and test dataframes containing columns f329 to f331 to analyze the null percentages and data ranges as per the instructions.



In [22]:
import numpy as np

def analyze_dataframe(df, name):
    print(f"Analyzing {name} DataFrame:")
    print("-" * 30)
    for col in df.columns:
        if df[col].dtype in [np.float64, np.int64]:
            null_percentage = df[col].isnull().sum() / len(df) * 100
            print(f"Column '{col}':")
            print(f"  Null Percentage: {null_percentage:.2f}%")
            # Check if there are non-null values before calculating min/max
            if df[col].notnull().sum() > 0:
                data_range = f"{df[col].min():.2f} - {df[col].max():.2f}"
                print(f"  Range: {data_range}")
            else:
                print("  Range: No non-null values to determine range.")
            print("-" * 20)
        else:
            print(f"Column '{col}':")
            print("  Not a numeric data type.")
            print("-" * 20)


analyze_dataframe(f329to331_train, "f329to331_train")
analyze_dataframe(f329to331_test, "f329to331_test")

Analyzing f329to331_train DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329':
  Not a numeric data type.
--------------------
Column 'f330':
  Not a numeric data type.
--------------------
Column 'f331':
  Not a numeric data type.
--------------------
Analyzing f329to331_test DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329':
  Not a numeric data type.
--------------------
Column 'f330':
  Not a numeric data type.
--------------------
Column 'f331':
  Not a numeric data type.
--------------------


**Reasoning**:
The previous analysis shows that the columns f329, f330, and f331 are not numeric. To perform numerical analysis like calculating ranges and imputing, these columns need to be converted to a numeric data type.



In [23]:
for col in ['f329', 'f330', 'f331']:
    f329to331_train[col] = pd.to_numeric(f329to331_train[col], errors='coerce')
    f329to331_test[col] = pd.to_numeric(f329to331_test[col], errors='coerce')

print("Columns converted to float in f329to331_train:")
print(f329to331_train.dtypes)

print("\nColumns converted to float in f329to331_test:")
print(f329to331_test.dtypes)

print("\nFirst few rows of f329to331_train after conversion:")
display(f329to331_train.head())

print("\nFirst few rows of f329to331_test after conversion:")
display(f329to331_test.head())

/tmp/ipython-input-23-2129940143.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f329to331_train[col] = pd.to_numeric(f329to331_train[col], errors='coerce')
/tmp/ipython-input-23-2129940143.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f329to331_test[col] = pd.to_numeric(f329to331_test[col], errors='coerce')


Columns converted to float in f329to331_train:
id2      object
f329    float64
f330    float64
f331    float64
dtype: object

Columns converted to float in f329to331_test:
id2      object
f329    float64
f330    float64
f331    float64
dtype: object

First few rows of f329to331_train after conversion:


id2  f329  f330  f331
0  1366776   NaN   NaN   NaN
1  1366776   NaN   NaN   NaN
2  1366776   NaN   NaN   NaN
3  1366776   NaN   NaN   NaN
4  1366776   NaN   NaN   NaN


First few rows of f329to331_test after conversion:


id2      f329      f330      f331
46756   1362907  0.844821  1.129103  0.994659
57819   1082599  1.074305  1.135387  1.110548
15390   1888466  0.822594  1.000000  1.000000
145730  1888971  0.928949  1.054698  1.020814
146085  1256369  0.823446  1.161529  1.014230

**Reasoning**:
Now that the relevant columns are numeric, I can call the previously defined `analyze_dataframe` function again to calculate and print the null percentages and data ranges for these columns in both dataframes.



In [24]:
analyze_dataframe(f329to331_train, "f329to331_train")
analyze_dataframe(f329to331_test, "f329to331_test")

Analyzing f329to331_train DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329':
  Null Percentage: 25.42%
  Range: 0.00 - 21.81
--------------------
Column 'f330':
  Null Percentage: 25.42%
  Range: 0.00 - 39.04
--------------------
Column 'f331':
  Null Percentage: 25.41%
  Range: 0.00 - 28.34
--------------------
Analyzing f329to331_test DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329':
  Null Percentage: 1.82%
  Range: 0.00 - 15.00
--------------------
Column 'f330':
  Null Percentage: 1.82%
  Range: 0.00 - 21.56
--------------------
Column 'f331':
  Null Percentage: 1.82%
  Range: 0.00 - 17.35
--------------------


## Summary:

### Data Analysis Key Findings

*   In the training dataset, the 'f329' column has a null percentage of 25.42%, with a range of 0.00 - 21.81. 'f330' has a null percentage of 25.42% and a range of 0.00 - 39.04. 'f331' has a null percentage of 25.41% and a range of 0.00 - 28.34.
*   In the test dataset, the 'f329' column has a null percentage of 1.82%, with a range of 0.00 - 15.00. 'f330' has a null percentage of 1.82% and a range of 0.00 - 21.56. 'f331' has a null percentage of 1.82% and a range of 0.00 - 17.35.
*   The columns 'f329', 'f330', and 'f331' have a significant percentage of null values in both datasets, although the percentages are much higher in the training data (around 25%) compared to the test data (around 1.8%).
*   The 'id2' column was confirmed to be non-numeric in both datasets.

### Insights or Next Steps

*   The notable difference in null percentages between the train and test datasets for 'f329', 'f330', and 'f331' should be investigated. This might indicate a data collection issue or a difference in the nature of the data between the two sets.
*   Imputation strategies should be considered for handling the missing values in these columns, taking into account the differing null percentages between train and test.
*   Further exploration of the 'id2' column's role and potential use as a categorical feature is recommended if relevant to the analysis goals.

In [25]:
# Select the columns to merge for train and test data
cols_to_merge = ['f329', 'f330', 'f331']

# Calculate the row-wise median and create a new column 'f329_to_f331_median'
train_data['f329_to_f331_median'] = train_data[cols_to_merge].median(axis=1)
test_data['f329_to_f331_median'] = test_data[cols_to_merge].median(axis=1)

print("New column 'f329_to_f331_median' created in train_data.")
print("New column 'f329_to_f331_median' created in test_data.")

# Display the first few rows with the new column
print("\nTrain data with new median column:")
display(train_data[['id2'] + cols_to_merge + ['f329_to_f331_median']].head())

print("\nTest data with new median column:")
display(test_data[['id2'] + cols_to_merge + ['f329_to_f331_median']].head())

TypeError: Cannot convert [[nan nan nan]
 [nan nan nan]
 [nan nan nan]
 ...
 ['1.1566483638519238' '0.9496393257653992' '1.0406703799233972']
 [nan nan nan]
 ['1.286202131048753' '0.9145929501922008' '1.0754816956117554']] to numeric

In [26]:
import pandas as pd

for col in ['f329', 'f330', 'f331']:
    train_data[col] = pd.to_numeric(train_data[col], errors='coerce')
    test_data[col] = pd.to_numeric(test_data[col], errors='coerce')

print("Columns converted to float in train_data:")
print(train_data[['f329', 'f330', 'f331']].dtypes)

print("\nColumns converted to float in test_data:")
print(test_data[['f329', 'f330', 'f331']].dtypes)

print("\nFirst few rows of train_data after conversion:")
display(train_data[['f329', 'f330', 'f331']].head())

print("\nFirst few rows of test_data after conversion:")
display(test_data[['f329', 'f330', 'f331']].head())

Columns converted to float in train_data:
f329    float64
f330    float64
f331    float64
dtype: object

Columns converted to float in test_data:
f329    float64
f330    float64
f331    float64
dtype: object

First few rows of train_data after conversion:


f329  f330  f331
0   NaN   NaN   NaN
1   NaN   NaN   NaN
2   NaN   NaN   NaN
3   NaN   NaN   NaN
4   NaN   NaN   NaN


First few rows of test_data after conversion:


f329      f330      f331
46756   0.844821  1.129103  0.994659
57819   1.074305  1.135387  1.110548
15390   0.822594  1.000000  1.000000
145730  0.928949  1.054698  1.020814
146085  0.823446  1.161529  1.014230

In [27]:
# Calculate the row-wise median and create a new column 'f329_to_f331_median'
cols_to_merge = ['f329', 'f330', 'f331']
train_data['f329_to_f331_median'] = train_data[cols_to_merge].median(axis=1)
test_data['f329_to_f331_median'] = test_data[cols_to_merge].median(axis=1)

print("New column 'f329_to_f331_median' created in train_data.")
print("New column 'f329_to_f331_median' created in test_data.")

# Display the first few rows with the new column
print("\nTrain data with new median column:")
display(train_data[['id2'] + cols_to_merge + ['f329_to_f331_median']].head())

print("\nTest data with new median column:")
display(test_data[['id2'] + cols_to_merge + ['f329_to_f331_median']].head())

New column 'f329_to_f331_median' created in train_data.
New column 'f329_to_f331_median' created in test_data.

Train data with new median column:


id2  f329  f330  f331  f329_to_f331_median
0  1366776   NaN   NaN   NaN                  NaN
1  1366776   NaN   NaN   NaN                  NaN
2  1366776   NaN   NaN   NaN                  NaN
3  1366776   NaN   NaN   NaN                  NaN
4  1366776   NaN   NaN   NaN                  NaN


Test data with new median column:


id2      f329      f330      f331  f329_to_f331_median
46756   1362907  0.844821  1.129103  0.994659             0.994659
57819   1082599  1.074305  1.135387  1.110548             1.110548
15390   1888466  0.822594  1.000000  1.000000             1.000000
145730  1888971  0.928949  1.054698  1.020814             1.020814
146085  1256369  0.823446  1.161529  1.014230             1.014230

In [28]:
# Analyze the new median column in train and test data
analyze_dataframe(train_data[['id2', 'f329_to_f331_median']], "train_data with f329_to_f331_median")
analyze_dataframe(test_data[['id2', 'f329_to_f331_median']], "test_data with f329_to_f331_median")

Analyzing train_data with f329_to_f331_median DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329_to_f331_median':
  Null Percentage: 25.41%
  Range: 0.00 - 3.50
--------------------
Analyzing test_data with f329_to_f331_median DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329_to_f331_median':
  Null Percentage: 1.82%
  Range: 0.00 - 3.50
--------------------


In [30]:
import pandas as pd

train_data = pd.read_parquet('/content/drive/MyDrive/AMEXChallenge/train_data.parquet')
test_data = pd.read_parquet('/content/drive/MyDrive/AMEXChallenge/test_data.parquet')

print("Original train and test data reloaded.")

# Display the columns to confirm
print("\nColumns in train_data after reloading:")
print(train_data.columns)

print("\nColumns in test_data after reloading:")
print(test_data.columns)

Original train and test data reloaded.

Columns in train_data after reloading:
Index(['id1', 'id2', 'id3', 'id4', 'id5', 'y', 'f1', 'f2', 'f3', 'f4',
       ...
       'f357', 'f358', 'f359', 'f360', 'f361', 'f362', 'f363', 'f364', 'f365',
       'f366'],
      dtype='object', length=372)

Columns in test_data after reloading:
Index(['id1', 'id2', 'id3', 'id4', 'id5', 'f1', 'f2', 'f3', 'f4', 'f5',
       ...
       'f357', 'f358', 'f359', 'f360', 'f361', 'f362', 'f363', 'f364', 'f365',
       'f366'],
      dtype='object', length=371)


In [31]:
cols_to_check = ['f329', 'f330', 'f331']

print("Checking for columns in train_data:")
for col in cols_to_check:
    if col in train_data.columns:
        print(f"Column '{col}' is present in train_data.")
    else:
        print(f"Column '{col}' is NOT present in train_data.")

print("\nChecking for columns in test_data:")
for col in cols_to_check:
    if col in test_data.columns:
        print(f"Column '{col}' is present in test_data.")
    else:
        print(f"Column '{col}' is NOT present in test_data.")

Checking for columns in train_data:
Column 'f329' is present in train_data.
Column 'f330' is present in train_data.
Column 'f331' is present in train_data.

Checking for columns in test_data:
Column 'f329' is present in test_data.
Column 'f330' is present in test_data.
Column 'f331' is present in test_data.


In [32]:
f329to331_train = train_data[['id2', 'f329', 'f330', 'f331']]
f329to331_test = test_data[['id2', 'f329', 'f330', 'f331']]

print("Train subset created successfully.")
display(f329to331_train.head())

print("\nTest subset created successfully.")
display(f329to331_test.head())

Train subset created successfully.


id2  f329  f330  f331
0  1366776  None  None  None
1  1366776  None  None  None
2  1366776  None  None  None
3  1366776  None  None  None
4  1366776  None  None  None


Test subset created successfully.


id2                f329                f330                f331
46756   1362907   0.844820822801228  1.1291026465763951  0.9946585915896966
57819   1082599   1.074304783146098   1.135386723678609  1.1105480081861696
15390   1888466  0.8225943721954622                 1.0                 1.0
145730  1888971  0.9289488395977646  1.0546975175685886  1.0208140575700744
146085  1256369  0.8234455057386793  1.1615293672707585   1.014230435062785

In [33]:
import pandas as pd
import numpy as np

# Convert columns to numeric
for col in ['f329', 'f330', 'f331']:
    f329to331_train[col] = pd.to_numeric(f329to331_train[col], errors='coerce')
    f329to331_test[col] = pd.to_numeric(f329to331_test[col], errors='coerce')

print("Columns converted to float in f329to331_train:")
print(f329to331_train.dtypes)

print("\nColumns converted to float in f329to331_test:")
print(f329to331_test.dtypes)

# Calculate the row-wise median and create a new column
cols_to_merge = ['f329', 'f330', 'f331']
f329to331_train['f329_to_f331_median'] = f329to331_train[cols_to_merge].median(axis=1)
f329to331_test['f329_to_f331_median'] = f329to331_test[cols_to_merge].median(axis=1)

print("\nNew column 'f329_to_f331_median' created in f329to331_train and f329to331_test.")

# Display the first few rows with the new column
print("\nTrain subset with new median column:")
display(f329to331_train.head())

print("\nTest subset with new median column:")
display(f329to331_test.head())

/tmp/ipython-input-33-3918683929.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f329to331_train[col] = pd.to_numeric(f329to331_train[col], errors='coerce')
/tmp/ipython-input-33-3918683929.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f329to331_test[col] = pd.to_numeric(f329to331_test[col], errors='coerce')


Columns converted to float in f329to331_train:
id2      object
f329    float64
f330    float64
f331    float64
dtype: object

Columns converted to float in f329to331_test:
id2      object
f329    float64
f330    float64
f331    float64
dtype: object

New column 'f329_to_f331_median' created in f329to331_train and f329to331_test.

Train subset with new median column:


/tmp/ipython-input-33-3918683929.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f329to331_train['f329_to_f331_median'] = f329to331_train[cols_to_merge].median(axis=1)
/tmp/ipython-input-33-3918683929.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f329to331_test['f329_to_f331_median'] = f329to331_test[cols_to_merge].median(axis=1)


id2  f329  f330  f331  f329_to_f331_median
0  1366776   NaN   NaN   NaN                  NaN
1  1366776   NaN   NaN   NaN                  NaN
2  1366776   NaN   NaN   NaN                  NaN
3  1366776   NaN   NaN   NaN                  NaN
4  1366776   NaN   NaN   NaN                  NaN


Test subset with new median column:


id2      f329      f330      f331  f329_to_f331_median
46756   1362907  0.844821  1.129103  0.994659             0.994659
57819   1082599  1.074305  1.135387  1.110548             1.110548
15390   1888466  0.822594  1.000000  1.000000             1.000000
145730  1888971  0.928949  1.054698  1.020814             1.020814
146085  1256369  0.823446  1.161529  1.014230             1.014230

In [34]:
# Drop the original f329, f330, and f331 columns
cols_to_drop = ['f329', 'f330', 'f331']

f329to331_train = f329to331_train.drop(columns=cols_to_drop)
f329to331_test = f329to331_test.drop(columns=cols_to_drop)

print(f"Columns {cols_to_drop} dropped from f329to331_train.")
print(f"Columns {cols_to_drop} dropped from f329to331_test.")

# Display the columns to verify
print("\nColumns in f329to331_train after dropping:")
print(f329to331_train.columns)

print("\nColumns in f329to331_test after dropping:")
print(f329to331_test.columns)

# Analyze the remaining columns
analyze_dataframe(f329to331_train, "f329to331_train after dropping original columns")
analyze_dataframe(f329to331_test, "f329to331_test after dropping original columns")

Columns ['f329', 'f330', 'f331'] dropped from f329to331_train.
Columns ['f329', 'f330', 'f331'] dropped from f329to331_test.

Columns in f329to331_train after dropping:
Index(['id2', 'f329_to_f331_median'], dtype='object')

Columns in f329to331_test after dropping:
Index(['id2', 'f329_to_f331_median'], dtype='object')
Analyzing f329to331_train after dropping original columns DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329_to_f331_median':
  Null Percentage: 25.41%
  Range: 0.00 - 3.50
--------------------
Analyzing f329to331_test after dropping original columns DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329_to_f331_median':
  Null Percentage: 1.82%
  Range: 0.00 - 3.50
--------------------


In [35]:
from sklearn.preprocessing import MinMaxScaler
import numpy as np

# Define the column to normalize
col_to_normalize = 'f329_to_f331_median'

# Initialize the MinMaxScaler
scaler = MinMaxScaler()

# Apply normalization
# Reshape the column to a 2D array as required by MinMaxScaler
f329to331_train[col_to_normalize] = scaler.fit_transform(f329to331_train[[col_to_normalize]])
f329to331_test[col_to_normalize] = scaler.transform(f329to331_test[[col_to_normalize]])

print(f"Column '{col_to_normalize}' normalized in both f329to331_train and f329to331_test.")

# Analyze the normalized column
analyze_dataframe(f329to331_train, "f329to331_train after normalization")
analyze_dataframe(f329to331_test, "f329to331_test after normalization")

print("\nFirst few rows of f329to331_train after normalization:")
display(f329to331_train.head())

print("\nFirst few rows of f329to331_test after normalization:")
display(f329to331_test.head())

Column 'f329_to_f331_median' normalized in both f329to331_train and f329to331_test.
Analyzing f329to331_train after normalization DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329_to_f331_median':
  Null Percentage: 25.41%
  Range: 0.00 - 1.00
--------------------
Analyzing f329to331_test after normalization DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f329_to_f331_median':
  Null Percentage: 1.82%
  Range: 0.00 - 1.00
--------------------

First few rows of f329to331_train after normalization:


id2  f329_to_f331_median
0  1366776                  NaN
1  1366776                  NaN
2  1366776                  NaN
3  1366776                  NaN
4  1366776                  NaN


First few rows of f329to331_test after normalization:


id2  f329_to_f331_median
46756   1362907             0.284317
57819   1082599             0.317443
15390   1888466             0.285843
145730  1888971             0.291793
146085  1256369             0.289911

In [36]:
# Define the column to impute
col_to_impute = 'f329_to_f331_median'

# Calculate the overall median for the column as a fallback
overall_median_train = f329to331_train[col_to_impute].median()
overall_median_test = f329to331_test[col_to_impute].median()

# Define a function to impute based on group median with overall median fallback
def impute_with_group_median(group, overall_median):
    return group.fillna(group.median()).fillna(overall_median)

# Apply the imputation to the training data
f329to331_train[col_to_impute] = f329to331_train.groupby('id2')[col_to_impute].transform(
    lambda x: impute_with_group_median(x, overall_median_train)
)

# Apply the imputation to the test data
f329to331_test[col_to_impute] = f329to331_test.groupby('id2')[col_to_impute].transform(
    lambda x: impute_with_group_median(x, overall_median_test)
)

print(f"Missing values in '{col_to_impute}' imputed based on group median with overall median fallback.")

print("\nFirst few rows of f329to331_train after imputation:")
display(f329to331_train.head())

print("\nFirst few rows of f329to331_test after imputation:")
display(f329to331_test.head())

# Verify if there are any remaining null values in the imputed column
print(f"\nNull value check after imputation for '{col_to_impute}' in f329to331_train:")
print(f329to331_train[col_to_impute].isnull().sum())

print(f"\nNull value check after imputation for '{col_to_impute}' in f329to331_test:")
print(f329to331_test[col_to_impute].isnull().sum())

Missing values in 'f329_to_f331_median' imputed based on group median with overall median fallback.

First few rows of f329to331_train after imputation:


id2  f329_to_f331_median
0  1366776             0.288756
1  1366776             0.288756
2  1366776             0.288756
3  1366776             0.288756
4  1366776             0.288756


First few rows of f329to331_test after imputation:


id2  f329_to_f331_median
46756   1362907             0.284317
57819   1082599             0.317443
15390   1888466             0.285843
145730  1888971             0.291793
146085  1256369             0.289911


Null value check after imputation for 'f329_to_f331_median' in f329to331_train:
0

Null value check after imputation for 'f329_to_f331_median' in f329to331_test:
0


In [37]:
import os
import pandas as pd

# Define the directory path
save_directory = '/content/drive/MyDrive/Transformed_Test_Data'

# Create the directory if it doesn't exist
os.makedirs(save_directory, exist_ok=True)

# Define the file path for saving the test DataFrame
file_path_save_test_329_331 = os.path.join(save_directory, 'f329to331_test_processed.parquet')

try:
    # Save the test DataFrame
    f329to331_test.to_parquet(file_path_save_test_329_331, index=False)
    print(f"Successfully saved the processed f329to331 test DataFrame to {file_path_save_test_329_331}")

except Exception as e:
    print(f"An error occurred while saving the DataFrame: {e}")

Successfully saved the processed f329to331 test DataFrame to /content/drive/MyDrive/Transformed_Test_Data/f329to331_test_processed.parquet


In [38]:
import os
import pandas as pd

# Define the directory path
save_directory = '/content/drive/MyDrive/Transformed_Train_Data'

# Create the directory if it doesn't exist
os.makedirs(save_directory, exist_ok=True)

# Define the file path for saving the train DataFrame
file_path_save_train_329_331 = os.path.join(save_directory, 'f329to331_train_processed.parquet')

try:
    # Save the train DataFrame
    f329to331_train.to_parquet(file_path_save_train_329_331, index=False)
    print(f"Successfully saved the processed f329to331 train DataFrame to {file_path_save_train_329_331}")

except Exception as e:
    print(f"An error occurred while saving the DataFrame: {e}")

Successfully saved the processed f329to331 train DataFrame to /content/drive/MyDrive/Transformed_Train_Data/f329to331_train_processed.parquet


In [39]:
f332to334_train = train_data[['id2', 'f332', 'f333', 'f334']]
f332to334_test = test_data[['id2', 'f332', 'f333', 'f334']]

print("Train subset created successfully.")
display(f332to334_train.head())

print("\nTest subset created successfully.")
display(f332to334_test.head())

Train subset created successfully.


id2 f332 f333 f334
0  1366776  1.0  0.0  0.0
1  1366776  1.0  0.0  0.0
2  1366776  1.0  0.0  0.0
3  1366776  1.0  0.0  0.0
4  1366776  1.0  0.0  0.0


Test subset created successfully.


id2 f332 f333 f334
46756   1362907  1.0  0.0  0.0
57819   1082599  0.0  1.0  0.0
15390   1888466  1.0  0.0  0.0
145730  1888971  1.0  0.0  0.0
146085  1256369  1.0  0.0  0.0

In [41]:
import pandas as pd

for col in ['f332', 'f333', 'f334']:
    f332to334_train[col] = pd.to_numeric(f332to334_train[col], errors='coerce')
    f332to334_test[col] = pd.to_numeric(f332to334_test[col], errors='coerce')

print("Columns converted to float in f332to334_train:")
print(f332to334_train.dtypes)

print("\nColumns converted to float in f332to334_test:")
print(f332to334_test.dtypes)

print("\nFirst few rows of f332to334_train after conversion:")
display(f332to334_train.head())

print("\nFirst few rows of f332to334_test after conversion:")
display(f332to334_test.head())

/tmp/ipython-input-41-684932012.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f332to334_train[col] = pd.to_numeric(f332to334_train[col], errors='coerce')
/tmp/ipython-input-41-684932012.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  f332to334_test[col] = pd.to_numeric(f332to334_test[col], errors='coerce')


Columns converted to float in f332to334_train:
id2      object
f332    float64
f333    float64
f334    float64
dtype: object

Columns converted to float in f332to334_test:
id2      object
f332    float64
f333    float64
f334    float64
dtype: object

First few rows of f332to334_train after conversion:


id2  f332  f333  f334
0  1366776   1.0   0.0   0.0
1  1366776   1.0   0.0   0.0
2  1366776   1.0   0.0   0.0
3  1366776   1.0   0.0   0.0
4  1366776   1.0   0.0   0.0


First few rows of f332to334_test after conversion:


id2  f332  f333  f334
46756   1362907   1.0   0.0   0.0
57819   1082599   0.0   1.0   0.0
15390   1888466   1.0   0.0   0.0
145730  1888971   1.0   0.0   0.0
146085  1256369   1.0   0.0   0.0

In [42]:
analyze_dataframe(f332to334_train, "f332to334_train after conversion")
analyze_dataframe(f332to334_test, "f332to334_test after conversion")

Analyzing f332to334_train after conversion DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f332':
  Null Percentage: 0.00%
  Range: 0.00 - 1.00
--------------------
Column 'f333':
  Null Percentage: 0.00%
  Range: 0.00 - 1.00
--------------------
Column 'f334':
  Null Percentage: 0.00%
  Range: 0.00 - 0.00
--------------------
Analyzing f332to334_test after conversion DataFrame:
------------------------------
Column 'id2':
  Not a numeric data type.
--------------------
Column 'f332':
  Null Percentage: 0.00%
  Range: 0.00 - 1.00
--------------------
Column 'f333':
  Null Percentage: 0.00%
  Range: 0.00 - 1.00
--------------------
Column 'f334':
  Null Percentage: 0.00%
  Range: 0.00 - 0.00
--------------------


In [43]:
import os
import pandas as pd

# Define the directory paths
save_directory_train = '/content/drive/MyDrive/Transformed_Train_Data'
save_directory_test = '/content/drive/MyDrive/Transformed_Test_Data'


# Create the directories if they don't exist
os.makedirs(save_directory_train, exist_ok=True)
os.makedirs(save_directory_test, exist_ok=True)

# Define the file paths for saving
file_path_save_train_332_334 = os.path.join(save_directory_train, 'f332to334_train_processed.parquet')
file_path_save_test_332_334 = os.path.join(save_directory_test, 'f332to334_test_processed.parquet')

try:
    # Save the train DataFrame
    f332to334_train.to_parquet(file_path_save_train_332_334, index=False)
    print(f"Successfully saved the processed f332to334 train DataFrame to {file_path_save_train_332_334}")

    # Save the test DataFrame
    f332to334_test.to_parquet(file_path_save_test_332_334, index=False)
    print(f"Successfully saved the processed f332to334 test DataFrame to {file_path_save_test_332_334}")

except Exception as e:
    print(f"An error occurred while saving the DataFrames: {e}")

Successfully saved the processed f332to334 train DataFrame to /content/drive/MyDrive/Transformed_Train_Data/f332to334_train_processed.parquet
Successfully saved the processed f332to334 test DataFrame to /content/drive/MyDrive/Transformed_Test_Data/f332to334_test_processed.parquet
